File: RoyalVNCObjCDemo/Source/Configuration/ConfigurationWindowController.m
#import "ConfigurationWindowController.h"
#import "ConfigurationViewController.h"

#import "ConnectionWindowController.h"

@import RoyalVNCKit;

@interface ConfigurationWindowController () <ConnectionWindowControllerDelegate>

@property BOOL didLoad;

@property (strong) ConfigurationViewController* configurationViewController;
@property (strong) NSMutableArray<ConnectionWindowController*>* connectionWindowControllers;

@property (weak) IBOutlet NSView* viewPlaceholderConfiguration;
@property (weak) IBOutlet NSButton* buttonConnect;

@end

@implementation ConfigurationWindowController

- (NSNibName)windowNibName {
	return @"ConfigurationWindow";
}

- (instancetype)init {
	self = [super initWithWindow:nil];
	
	if (self) {
		VNCConnectionSettings* settings = [VNCConnectionSettings fromUserDefaults];
		
		self.configurationViewController = [[ConfigurationViewController alloc] initWithSettings:settings];
		self.connectionWindowControllers = [NSMutableArray array];
	}
	
	return self;
}

- (void)windowDidLoad {
    [super windowDidLoad];
    
	if (self.didLoad) {
		return;
	}
	
	self.didLoad = YES;
	
	[self configureUI];
}

- (void)configureUI {
	NSView* configView = self.configurationViewController.view;
	configView.frame = self.viewPlaceholderConfiguration.bounds;
	configView.autoresizingMask = NSViewMinXMargin | NSViewMaxXMargin | NSViewMinYMargin | NSViewMaxYMargin | NSViewWidthSizable | NSViewHeightSizable;

	[self.viewPlaceholderConfiguration addSubview:configView];

	[self.window recalculateKeyViewLoop];
	[self.window makeFirstResponder:self.configurationViewController];
}

- (IBAction)buttonConnect_action:(id)sender {
	[self connect];
}

- (void)saveSettings {
	[self.configurationViewController.settings saveToUserDefaults];
}

- (NSNumber*)colorDepthOfActiveConnection {
	VNCConnection* activeConnection = self.activeConnection;
	
	if (!activeConnection) {
		return nil;
	}
	
	VNCFramebuffer* framebuffer = activeConnection.framebuffer;
	
	if (!framebuffer) {
		return nil;
	}
	
	VNCColorDepth colorDepth = framebuffer.colorDepth;
	
	return @(colorDepth);
}

- (void)connect {
	NSWindow* window = self.window;
	
	if (!window) {
		return;
	}
	
	VNCConnectionSettings* settings = self.configurationViewController.settings;
	[settings saveToUserDefaults];
	
	if ([VNCInputModeUtils inputModeRequiresAccessibilityPermissions:settings.inputMode] &&
		!VNCAccessibilityUtils.hasAccessibilityPermissions) {
		// Requires accessibility permissions but don't have them right now so ask user
		
		NSAlert* alert = [NSAlert new];
		alert.messageText = @"Accessibility Permissions";
		alert.informativeText = @"Accessibility Permissions are required when the input mode is set to \"Forward all keyboard shortcuts and hot keys\". To continue, please open System Settings and grant the app accessibility permissions.";
		
		[alert addButtonWithTitle:@"Open System Settings"];
		[alert addButtonWithTitle:@"Continue without permissions"];
		[alert addButtonWithTitle:@"Cancel"];
		
		__weak ConfigurationWindowController* weakSelf = self;
		
		[alert beginSheetModalForWindow:window completionHandler:^(NSModalResponse response) {
			switch (response) {
				case NSAlertFirstButtonReturn:
					[VNCAccessibilityUtils openAccessibilityPermissionsPreferencePane];
					
					break;
				case NSAlertSecondButtonReturn:
					[weakSelf connectWithSettings:settings];
					
					break;
				default:
					return;
			}
		}];
	} else {
		[self connectWithSettings:settings];
	}
}

- (void)connectWithSettings:(VNCConnectionSettings*)settings {
	ConnectionWindowController* connectionWindowController = [[ConnectionWindowController alloc] initWithSettings:settings];
	connectionWindowController.delegate = self;
	
	[self.connectionWindowControllers addObject:connectionWindowController];
	
	[connectionWindowController showWindow:self];
	[connectionWindowController connect];
}

- (nullable VNCConnection*)activeConnection {
	ConnectionWindowController* activeConnectionWindowController = self.activeConnectionWindowController;
	
	if (!activeConnectionWindowController) {
		return nil;
	}
	
	return activeConnectionWindowController.connection;
}

- (nullable ConnectionWindowController*)activeConnectionWindowController {
	NSWindow* keyWindow = NSApplication.sharedApplication.keyWindow;
	
	if (!keyWindow) {
		return nil;
	}
	
	ConnectionWindowController* keyConnectionWindowController = nil;
	
	for (ConnectionWindowController* connectionWindowController in self.connectionWindowControllers) {
		if (connectionWindowController.window == keyWindow) {
			keyConnectionWindowController = connectionWindowController;
			
			break;
		}
	}
	
	return keyConnectionWindowController;
}

#pragma mark - ConnectionWindowControllerDelegate

- (void)connectionWindowControllerDidClose:(ConnectionWindowController*)connectionWindowController {
	connectionWindowController.delegate = nil;
	
	[self.connectionWindowControllers removeObject:connectionWindowController];
}

@end

File: RoyalVNCDemo/Source/Configuration/ConfigurationWindowController.swift
import Foundation
import AppKit

import RoyalVNCKit

class ConfigurationWindowController: NSWindowController {
	@IBOutlet private weak var viewPlaceholderConfiguration: NSView!
	@IBOutlet private weak var buttonConnect: NSButton!
	
	override var windowNibName: NSNib.Name? { "ConfigurationWindow" }
	
	private var didLoad = false
	
	private let configurationViewController = ConfigurationViewController(settings: .fromUserDefaults())
	private var connectionWindowControllers = [ConnectionWindowController]()
	
	// MARK: - UI Events
	override func windowDidLoad() {
		super.windowDidLoad()
		
		guard !didLoad else { return }
		didLoad = true
		
		configureUI()
	}
	
	@IBAction private func buttonConnect_action(_ sender: Any) {
		connect()
	}
}

extension ConfigurationWindowController {
	func saveSettings() {
		configurationViewController.settings.saveToUserDefaults()
	}
	
	var colorDepthOfActiveConnection: VNCConnection.Settings.ColorDepth? {
		guard let activeConnection = activeConnection else {
			return nil
		}
		
		guard let framebuffer = activeConnection.framebuffer else {
			return nil
		}
		
		let colorDepth = framebuffer.colorDepth
		
		return colorDepth
	}
}

private extension ConfigurationWindowController {
	func configureUI() {
		let configView = configurationViewController.view
		configView.frame = viewPlaceholderConfiguration.bounds
		configView.autoresizingMask = [ .minXMargin, .maxXMargin, .minYMargin, .maxYMargin, .width, .height ]
		
		viewPlaceholderConfiguration.addSubview(configView)
		
		window?.recalculateKeyViewLoop()
		window?.makeFirstResponder(configurationViewController)
	}
	
	func connect() {
        guard let window = window else { return }
        
		let settings = configurationViewController.settings
		settings.saveToUserDefaults()
        
		if settings.inputMode.requiresAccessibilityPermissions &&
			!VNCAccessibilityUtils.hasAccessibilityPermissions {
			// Requires accessibility permissions but don't have them right now so ask user
			
            let alert = NSAlert()
            alert.messageText = "Accessibility Permissions"
            alert.informativeText = "Accessibility Permissions are required when the input mode is set to \"Forward all keyboard shortcuts and hot keys\". To continue, please open System Settings and grant the app accessibility permissions."
            
            alert.addButton(withTitle: "Open System Settings")
			alert.addButton(withTitle: "Continue without permissions")
            alert.addButton(withTitle: "Cancel")
            
            alert.beginSheetModal(for: window) { [weak self] response in
				switch response {
					case .alertFirstButtonReturn:
						VNCAccessibilityUtils.openAccessibilityPermissionsPreferencePane()
					case .alertSecondButtonReturn:
						self?.connect(settings: settings)
					default:
						return
				}
                
            }
		} else {
			connect(settings: settings)
		}
	}
    
    func connect(settings: VNCConnection.Settings) {
        let connectionWindowController = ConnectionWindowController(settings: settings)
        connectionWindowController.delegate = self
        
        connectionWindowControllers.append(connectionWindowController)
        
        connectionWindowController.showWindow(self)
        connectionWindowController.connect()
    }
	
	var activeConnection: VNCConnection? {
		return activeConnectionWindowController?.connection
	}
	
	var activeConnectionWindowController: ConnectionWindowController? {
		guard let keyWindow = NSApplication.shared.keyWindow else {
			return nil
		}
		
		guard let keyConnectionWindowController = connectionWindowControllers.first(where: { $0.window == keyWindow }) else {
			return nil
		}
		
		return keyConnectionWindowController
	}
}

extension ConfigurationWindowController: ConnectionWindowControllerDelegate {
	func connectionWindowControllerDidClose(_ connectionWindowController: ConnectionWindowController) {
		connectionWindowController.delegate = nil
		
		connectionWindowControllers.removeAll { $0 == connectionWindowController }
	}
}
